blockchain/store_test.go

```diff
@@ -0,0 +1,267 @@
+
+import (
+	"fmt"
+	"io/ioutil"
+	"runtime"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/tendermint/go-wire/data"
+	"github.com/tendermint/tendermint/blockchain"
+	"github.com/tendermint/tendermint/types"
+	"github.com/tendermint/tmlibs/db"
+)
+
+func TestLoadBlockStoreStateJSON(t *testing.T) {
+	db := db.NewMemDB()
+
+	bsj := &blockchain.BlockStoreStateJSON{Height: 1000}
+	bsj.Save(db)
+
+	retrBSJ := blockchain.LoadBlockStoreStateJSON(db)
+
+	assert.Equal(t, *bsj, retrBSJ, "expected the retrieved DBs to match")
+}
+
+var blockStoreKey = []byte("blockStore")
+
+func TestNewBlockStore(t *testing.T) {
+	db := db.NewMemDB()
+	db.Set(blockStoreKey, []byte(`{"height": 10000}`))
+	bs := blockchain.NewBlockStore(db)
+	assert.Equal(t, bs.Height(), 10000, "failed to properly parse blockstore")
+
+	panicCausers := [][]byte{
+		[]byte("artful-doger"),
+		[]byte(""),
+		[]byte(" "),
+	}
+
+	for i, panicCauser := range panicCausers {
+		// Expecting a panic here on trying to parse an invalid blockStore
+		_, _, err := doFn(func() (interface{}, error) {
+			db.Set(blockStoreKey, panicCauser)
+			_ = blockchain.NewBlockStore(db)
+			return nil, nil
+		})
+		assert.NotNil(t, err, "#%d panicCauser: %q expected a panic", i, panicCauser)
+		// Note: We can't check for the exact substring because
+		// recovered stack traces as of Sept 2017 seem to
+		// mask the message. See https://play.golang.org/p/W9NTc4Fstr
+		// TODO: Perhaps file an error on the Go project.
+	}
+
+	db.Set(blockStoreKey, nil)
+	bs = blockchain.NewBlockStore(db)
+	assert.Equal(t, bs.Height(), 0, "expecting nil bytes to be unmarshaled alright")
+}
+
+func TestBlockStoreGetReader(t *testing.T) {
+	db := db.NewMemDB()
+	// Initial setup
+	db.Set([]byte("Foo"), []byte("Bar"))
+	db.Set([]byte("Foo1"), nil)
+
+	bs := blockchain.NewBlockStore(db)
+
+	tests := [...]struct {
+		key  []byte
+		want []byte
+	}{
+		0: {key: []byte("Foo"), want: []byte("Bar")},
+		1: {key: []byte("KnoxNonExistent"), want: nil},
+		2: {key: []byte("Foo1"), want: nil},
+	}
+
+	for i, tt := range tests {
+		r := bs.GetReader(tt.key)
+		if r == nil {
+			assert.Nil(t, tt.want, "#%d: expected a non-nil reader", i)
+			continue
+		}
+		slurp, err := ioutil.ReadAll(r)
+		if err != nil {
+			t.Errorf("#%d: unexpected Read err: %v", i, err)
+		} else {
+			assert.Equal(t, slurp, tt.want, "#%d: mismatch", i)
+		}
+	}
+}
+
+func freshBlockStore() (*blockchain.BlockStore, db.DB) {
+	db := db.NewMemDB()
+	return blockchain.NewBlockStore(db), db
+}
+
+func TestBlockStoreSaveLoadBlock(t *testing.T) {
+	bs, _ := freshBlockStore()
+	noBlockHeights := []int{0, -1, 100, 1000, 2}
+	for i, height := range noBlockHeights {
+		if g := bs.LoadBlock(height); g != nil {
+			t.Errorf("#%d: height(%d) got a block; want nil", i, height)
+		}
+	}
+
+	// Setup, test data
+	part2 := &types.Part{Index: 1, Bytes: data.Bytes([]byte{
+		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x01, 0x01, 0x01, 0x00,
```

Comment:
how did you generate these? or it's just random bytes?

Reply:
So I generated them by
```go
package main

import (
	"bytes"
	"log"

	"github.com/tendermint/go-wire"
	"github.com/tendermint/tendermint/types"
)

func main() {
	buf := new(bytes.Buffer)
	var n int
	var err error

	block := &types.Block{
		Header: &types.Header{
			Height:  1,
			NumTxs:  100,
			ChainID: "block_test",
		},
		LastCommit: &types.Commit{Precommits: []*types.Vote{{Height: 10}}},
	}
	wire.WriteBinary(block, buf, &n, &err)
	log.Printf("% x", buf.Bytes())
}
```
and then split the bytes over two parts.

However, I didn't include the generation code in the test, lest it caused confusion.

Reply:
But if something changes, how a next person supposed to change this stuff? Maybe we should at least add a comment?

Reply:
Let's include the generation code as well and ideally generate them on each test run.

Reply:
Roger that, I'll add that comment in in a second. @adrianbrink am not too sure about regenerating it on each test run as that makes the test harder to grok, however a note on how to generate the parts bytes.